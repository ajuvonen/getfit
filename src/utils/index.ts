import {unref} from 'vue';
import {helpers} from '@vuelidate/validators';
import type {ErrorObject} from '@vuelidate/core';
import {v4 as uuid} from 'uuid';
import {clone} from 'remeda';
import {Intensity, type ScheduleSettings, type Training} from '@/types';
import {ACTIVITIES, COLORS} from '@/constants';

export const getIcon = (activityName: string) => `$${activityName}`;

export const getIntensityColor = (intensity: Intensity, alpha: number = 1) => {
  const hexAlpha = alpha < 1 ? Math.round(alpha * 255).toString(16) : '';
  switch (intensity) {
    case Intensity.NORMAL:
      return COLORS.intensityNormal + hexAlpha;
    case Intensity.DEMANDING:
      return COLORS.intensityDemanding + hexAlpha;
    case Intensity.HEAVY:
      return COLORS.intensityHeavy + hexAlpha;
    default:
      return COLORS.intensityLight + hexAlpha;
  }
};

export const getEmptySettings = (
  initialSettings: Partial<ScheduleSettings> = {},
): ScheduleSettings => ({
  name: '',
  startsOnSunday: false,
  startDate: null,
  actualWeekNumbering: false,
  availableActivities: clone(ACTIVITIES),
  defaultStartTime: {
    hours: 12,
    minutes: 0,
    seconds: 0,
  },
  defaultDuration: 1,
  defaultUnitOfDuration: 'h',
  darkMode: 'auto',
  decoratedCards: true,
  ...initialSettings,
});

export const getEmptyTraining = (initialTraining: Partial<Training> = {}): Training => ({
  id: uuid(),
  weekId: uuid(),
  activity: '',
  dayIndex: 0,
  title: '',
  instructions: '',
  duration: 1,
  unitOfDuration: 'h',
  intensity: Intensity.NORMAL,
  location: '',
  completed: false,
  rating: null,
  ...initialTraining,
});

export const decimalRegex = helpers.regex(/^\d+(.(00?|25|50?|75))?$/);

export const getValidationErrors = (errors: ErrorObject[]) =>
  errors.map((error) => unref(error.$message));

export const isDurationTime = (value: string) => ['h', 'm'].includes(value);
