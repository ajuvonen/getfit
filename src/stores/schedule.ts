import {defineStore} from 'pinia';
import {useStorage} from '@vueuse/core';
import {v4 as uuidv4} from 'uuid';
import {clone, lensProp, over} from 'ramda';
import {DateTime} from 'luxon';
import type {ScheduleSettings, Training, Week} from '@/types';
import {ACTIVITIES} from '@/constants';
import {roundNearestQuarter} from '@/utils';
import {computed, watch} from 'vue';

const getEmptySchedule = (): ScheduleSettings => ({
  name: '',
  startsOnSunday: false,
  startDate: null,
  actualWeekNumbering: false,
  availableActivities: ACTIVITIES.map(({value}) => value),
  defaultStartTime: {
    hours: 12,
    minutes: 0,
    seconds: 0,
  },
  defaultDuration: 1,
  unitOfTime: 'h',
});

export const useScheduleStore = defineStore('schedule', () => {
  // State refs
  const settings = useStorage('getfit-settings', getEmptySchedule(), localStorage, {
    mergeDefaults: true,
  });

  const weeks = useStorage('getfit-schedule', [] as Week[], localStorage, {mergeDefaults: true});

  // Computed getters
  const getTargetWeekAndTraining = computed(() => (weekId: string, trainingId?: string) => {
    const targetWeek = weeks.value.find(({id}) => id === weekId) as Week;
    const targetTraining = trainingId
      ? targetWeek.trainings.find(({id}) => id === trainingId)
      : undefined;
    return [targetWeek, targetTraining] as [Week, Training | undefined];
  });

  // Actions
  const addWeek = () => {
    weeks.value.push({
      id: uuidv4(),
      trainings: [],
    });
  };

  const deleteWeek = (weekId: string) => {
    weeks.value = weeks.value.filter(({id}) => id !== weekId);
  };

  const copyWeek = (weekId: string) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    const newWeekId = uuidv4();
    const cloneWeek: Week = {
      id: newWeekId,
      trainings: targetWeek.trainings.map((training) => ({
        ...training,
        id: uuidv4(),
        weekId: newWeekId,
      })),
    };
    weeks.value.push(cloneWeek);
  };

  const addOrEditTraining = (training: Training) => {
    const [targetWeek] = getTargetWeekAndTraining.value(training.weekId);
    const targetIndex = targetWeek.trainings.findIndex(({id}) => id === training.id);
    if (targetIndex >= 0) {
      targetWeek.trainings[targetIndex] = training;
    } else {
      targetWeek.trainings.push(training);
    }
  };

  const deleteTraining = (training: Training) => {
    const [targetWeek] = getTargetWeekAndTraining.value(training.weekId);
    targetWeek.trainings = targetWeek.trainings.filter(({id}) => id !== training.id);
  };

  const moveTraining = (training: Training, newWeekId: string, dayIndex: number) => {
    const [originalWeek, targetTraining] = getTargetWeekAndTraining.value(
      training.weekId,
      training.id,
    ) as [Week, Training];
    const [targetWeek] = getTargetWeekAndTraining.value(newWeekId);
    originalWeek.trainings = originalWeek.trainings.filter(({id}) => id !== targetTraining.id);
    targetTraining.weekId = newWeekId;
    targetTraining.dayIndex = dayIndex;
    targetWeek.trainings.push(targetTraining);
  };

  const reorderTrainings = (weekId: string, trainings: Training[]) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    targetWeek.trainings = trainings;
  };

  const copyTraining = (training: Training, weekId: string, dayIndex: number) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    const clonedTraining: Training = {
      ...clone(training),
      id: uuidv4(),
      weekId,
      dayIndex,
    };
    targetWeek.trainings.push(clonedTraining);
  };

  // Watchers
  watch(
    () => settings.value.unitOfTime,
    (newValue, oldValue) => {
      if (newValue && newValue !== oldValue) {
        const multiplier = newValue === 'h' ? 1 / 60 : 60;
        const precision = newValue === 'h' ? 2 : 0;
        settings.value.defaultDuration = roundNearestQuarter(
          settings.value.defaultDuration * multiplier,
          precision,
        );
        weeks.value = weeks.value.map((week) => ({
          ...week,
          trainings: week.trainings.map(
            over(lensProp('duration'), (duration) =>
              roundNearestQuarter(duration * multiplier, precision),
            ),
          ),
        }));
      }
    },
  );

  watch(
    () => settings.value.startsOnSunday,
    (newValue) => {
      const parsedValue = newValue || false;
      if (settings.value.startDate) {
        const date = DateTime.fromJSDate(settings.value.startDate);
        if (parsedValue) {
          settings.value.startDate = date.minus({days: 1}).toJSDate();
        } else {
          settings.value.startDate = date.plus({days: 1}).toJSDate();
        }
      }
    },
  );

  // Reset
  const $reset = () => {
    settings.value = getEmptySchedule();
    weeks.value = [];
  };

  return {
    settings,
    weeks,
    getTargetWeekAndTraining,
    addWeek,
    deleteWeek,
    copyWeek,
    addOrEditTraining,
    deleteTraining,
    moveTraining,
    reorderTrainings,
    copyTraining,
    $reset,
  };
});
