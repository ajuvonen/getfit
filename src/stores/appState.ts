import {defineStore} from 'pinia';
import {v4 as uuidv4} from 'uuid';
import type {Training} from '@/types';
import {clone} from 'ramda';
import {useScheduleStore} from './schedule';

export const useAppStateStore = defineStore('appState', {
  state: () => ({
    summaryShown: [] as string[],
    trainingDialogOpen: false,
    trainingData: {} as Training,
  }),
  getters: {
    isSummaryShown: (state) => (trainingId: string) => state.summaryShown.includes(trainingId),
  },
  actions: {
    openNewTrainingDialog(weekId: string, dayIndex: number) {
      this.trainingData = {
        id: uuidv4(),
        weekId,
        dayIndex,
        duration: 0,
        activity: '',
        description: '',
        title: '',
        intensity: 0,
        completionSummary: '',
        location: '',
      };
      this.trainingDialogOpen = true;
    },
    openEditTrainingDialog(training: Training) {
      const scheduleStore = useScheduleStore();
      const clonedTraining = clone(training);
      if (
        clonedTraining.activity &&
        !scheduleStore.schedule.availableActivities.includes(clonedTraining.activity)
      ) {
        clonedTraining.activity = '';
      }
      this.trainingData = clonedTraining;
      this.trainingDialogOpen = true;
    },
    closeEditTrainingDialog() {
      this.trainingDialogOpen = false;
    },
    toggleSummaryShown(trainingId: string) {
      if (this.isSummaryShown(trainingId)) {
        this.summaryShown = this.summaryShown.filter((id) => id !== trainingId);
      } else {
        this.summaryShown.push(trainingId);
      }
    },
  },
});
