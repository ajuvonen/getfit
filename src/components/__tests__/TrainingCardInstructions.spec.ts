import {mount} from '@vue/test-utils';
import {describe, it, expect, beforeEach} from 'vitest';
import {v4 as uuid} from 'uuid';  
import {getEmptyTraining} from '@/utils';
import { useAppStateStore } from '@/stores/appState';
import TrainingCardInstructions from '@/components/TrainingCardInstructions.vue';

describe('TrainingCardInstructions', () => {
  let appStateStore: ReturnType<typeof useAppStateStore>;

  beforeEach(() => {
    appStateStore = useAppStateStore();
  });

  it('mounts', () => {
    const description = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit';
    const wrapper = mount(TrainingCardInstructions, {
      props: {
        training: getEmptyTraining({description}),
        show: true,
      },
    });
    expect(wrapper.find('.training-card__instructions').exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('remains hidden when show is false', () => {
    const wrapper = mount(TrainingCardInstructions, {
      props: {
        training: getEmptyTraining(),
        show: false,
      },
    });
    expect(wrapper.find('.training-card__instructions').exists()).toBe(false);
  });

  it('closes when clicked', async () => {
    const trainingId = uuid();
    const wrapper = mount(TrainingCardInstructions, {
      props: {
        training: getEmptyTraining({id: trainingId}),
        show: true,
      },
    });
    expect(wrapper.find('.training-card__instructions').exists()).toBe(true);
    await wrapper.find('.training-card__instructions__close-button').trigger('click');
    expect(appStateStore.toggleShowDescription).toHaveBeenCalledWith(trainingId);
  })
});
