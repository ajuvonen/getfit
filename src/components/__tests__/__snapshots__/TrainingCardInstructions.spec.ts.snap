// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrainingCardInstructions > mounts 1`] = `
"<transition-stub data-v-dc9f9ee2="" name="expand-transition" appear="false" persisted="false" css="true">
  <div data-v-dc9f9ee2="" class="v-card v-theme--customLight v-card--density-default v-card--variant-elevated d-flex flex-column" style="background-color: rgb(255, 217, 140); color: rgb(33, 33, 33); caret-color: #000; height: 100%;">
    <!---->
    <div class="v-card__loader">
      <div class="v-progress-linear v-theme--customLight v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
        <!---->
        <div class="v-progress-linear__background"></div>
        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
          <div class="v-progress-linear__indeterminate">
            <div class="v-progress-linear__indeterminate long"></div>
            <div class="v-progress-linear__indeterminate short"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-dc9f9ee2="" class="v-card-text">
      <div data-v-dc9f9ee2="" class="training-card__instructions">Lorem ipsum dolor sit amet, consectetur adipiscing elit</div>
    </div>
    <div data-v-dc9f9ee2="" class="v-card-actions justify-end"><button data-v-dc9f9ee2="" type="button" class="v-btn v-btn--slim v-theme--customLight bg-transparent v-btn--density-default v-btn--size-default v-btn--variant-flat training-card__close-instructions-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="v-icon notranslate v-theme--customLight v-icon--size-default" aria-hidden="true"><svg class="v-icon__svg" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true"><path d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"></path></svg></i></span><span class="v-btn__content" data-no-activator="">Close</span>
        <!---->
        <!---->
      </button></div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</transition-stub>"
`;
